SU2_CFD/src/solvers/CNEMONSSolver.cpp

```diff
@@ -977,149 +952,162 @@ void CNEMONSSolver::BC_Smoluchowski_Maxwell(CGeometry *geometry,
-      }
+    su2double Mass = 0.0;
+    for (auto iSpecies=0u; iSpecies<nSpecies; iSpecies++)
+      Mass += Vi[iSpecies]*Ms[iSpecies];
+    su2double Cptr = rhoCvtr + Ru/Mass;
+    su2double tmp1 = Cptr*(Eddy_Visc/Prandtl_Turb);
+    su2double scl  = tmp1/ktr;
+    ktr += Cptr*(Eddy_Visc/Prandtl_Turb);
+    kve  = kve*(1.0+scl);
 
-      /*--- Apply to the linear system ---*/
-      Res_Visc[nSpecies+nDim]   = ((ktr*(Ti-Tj)    + kve*(Tvei-Tvej)) +
-                                   (ktr*(Tslip-Ti) + kve*(Tslip_ve-Tvei))*C)*Area/dij;
-      Res_Visc[nSpecies+nDim+1] = (kve*(Tvei-Tvej) + kve*(Tslip_ve-Tvei)*C)*Area/dij;
+    /*--- Retrieve Primitive Gradients ---*/
+    const auto Grad_PrimVar = nodes->GetGradient_Primitive(iPoint);
 
-      LinSysRes.SubtractBlock(iPoint, Res_Visc);
+    /*--- Calculate specific gas constant --- */
+    su2double GasConstant = FluidModel->ComputeGasConstant();
+
+    /*--- Calculate temperature gradients normal to surface---*/ //Doubt about minus sign
+    su2double dTn = GeometryToolbox::DotProduct(nDim, Grad_PrimVar[T_INDEX], UnitNormal);
+    su2double dTven = GeometryToolbox::DotProduct(nDim, Grad_PrimVar[TVE_INDEX], UnitNormal);
+
+    /*--- Calculate molecular mean free path ---*/
+    su2double Lambda = Viscosity/Density*sqrt(PI_NUMBER/(2.0*GasConstant*Ti));
+
+    /*--- Calculate Temperature Slip ---*/
+    su2double Tslip = ((2.0-TAC)/TAC)*2.0*Gamma/(Gamma+1.0)/Prandtl_Lam*Lambda*dTn+Twall;
+
+    su2double Tslip_ve = Twall;
+    if (dTven !=0) Tslip_ve = (Tslip-Twall)*(kve*rhoCvtr/dTn)/(ktr*rhoCvve/dTven)+Twall;
+
+    /*--- Calculate temperature gradients tangent to surface ---*/
+    su2double Vector_Tangent_dT[MAXNDIM] = {0.0}, Vector_Tangent_dTve[MAXNDIM] = {0.0};
+    for (auto iDim = 0u; iDim < nDim; iDim++) {
+      Vector_Tangent_dT[iDim]   = Grad_PrimVar[T_INDEX][iDim] - dTn * UnitNormal[iDim];
+      Vector_Tangent_dTve[iDim] = Grad_PrimVar[TVE_INDEX][iDim] - dTven * UnitNormal[iDim];
     }
+
+    /*--- Calculate Heatflux tangent to surface ---*/
+    su2double Vector_Tangent_HF[MAXNDIM] = {0.0};
+    for (auto iDim = 0u; iDim < nDim; iDim++)
+      Vector_Tangent_HF[iDim] = -ktr*Vector_Tangent_dT[iDim]-kve*Vector_Tangent_dTve[iDim];
+
+    /*--- Initialize viscous residual to zero ---*/
+    for (auto iVar = 0u; iVar < nVar; iVar ++) Res_Visc[iVar] = 0.0;
+
+    /*--- Compute wall shear stress (using the stress tensor) ---*/
+    su2double TauElem[MAXNDIM] = {0.0}, TauTangent[MAXNDIM] = {0.0};
+    su2double Tau[MAXNDIM][MAXNDIM] = {{0.0}};
+    CNumerics::ComputeStressTensor(nDim, Tau, Grad_PrimVar+VEL_INDEX, Viscosity);
+    for (auto iDim = 0u; iDim < nDim; iDim++)
+      TauElem[iDim] = GeometryToolbox::DotProduct(nDim, Tau[iDim], UnitNormal);
+
+    su2double TauNormal = GeometryToolbox::DotProduct(nDim, TauElem, UnitNormal);
+
+    for (auto iDim = 0u; iDim < nDim; iDim++)
+      TauTangent[iDim] = TauElem[iDim] - TauNormal * UnitNormal[iDim];
```

Comment:
I think I put a function in the toolbox for this particular pattern, tangent component or projection or something.

Reply:
Done!